static/app/components/core/slider/index.chonk.tsx

```diff
@@ -0,0 +1,317 @@
+import {
+  type CSSProperties,
+  useCallback,
+  useEffect,
+  useMemo,
+  useRef,
+  useState,
+} from 'react';
+import type {DO_NOT_USE_ChonkTheme} from '@emotion/react';
+import {css} from '@emotion/react';
+import styled from '@emotion/styled';
+import {mergeRefs} from '@react-aria/utils';
+
+import {space} from 'sentry/styles/space';
+
+import type {SliderProps} from './index';
+
+const passthrough = (n: number | '') => n;
+
+export function Slider({
+  formatLabel = passthrough,
+  ref,
+  ...props
+}: SliderProps & {ref: React.Ref<HTMLInputElement>}) {
+  const inputRef = useRef<HTMLInputElement>(null);
+  const refs = mergeRefs(ref, inputRef);
+  const step = toNumber(props.step ?? -1);
+  const {value, min, max} = resolveMinMaxValue(props);
+  const [label, setLabel] = useState(value);
+  const initialProgress = getProgress(value, min, max);
+  const filledSteps = useMemo(() => (step === -1 ? -1 : label / step), [step, label]);
+
+  const handleChange = useCallback(
+    (event: Event) => {
+      const input = event.target as HTMLInputElement;
+      const {valueAsNumber, min: nativeMin, max: nativeMax} = input;
+      setLabel(valueAsNumber);
+      const progress = getProgress(valueAsNumber, nativeMin || 0, nativeMax || 100);
+      input.parentElement?.style.setProperty('--p', `${progress.toFixed(0)}%`);
+    },
+    [setLabel]
+  );
+
+  useEffect(() => {
+    const input: HTMLInputElement | null = inputRef.current;
+    if (!input) {
+      return;
+    }
+    input.addEventListener('input', handleChange);
+    // eslint-disable-next-line consistent-return
+    return () => input.removeEventListener('input', handleChange);
+  });
```

Comment:
This effect will be creating a lot of callbacks as the state value changes - a callback ref would be perfect here, this way you can still hook into mount and unmount while preserving a single change handler

Reply:
callback ref works with react19 now since we can return cleanups 🎉 , so all in all, we’d get:

```
const inputRef = useCallback((input: HTMLInputElement | null) => {
  if (!input) {
    return undefined;
  }

  const handleChange = () => {
    const {valueAsNumber, min: nativeMin, max: nativeMax} = input;
    setLabel(valueAsNumber);
    const progress = getProgress(valueAsNumber, nativeMin || 0, nativeMax || 100);
    input.parentElement?.style.setProperty('--p', `${progress.toFixed(0)}%`);
  };

  input.addEventListener('input', handleChange);

  return () => input.removeEventListener('input', handleChange);
}, []);
```

which is pretty nice